monads/monads/test/Spec.hs

```diff
@@ -0,0 +1,113 @@
+import Hedgehog
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import Test.Tasty
+import Test.Tasty.Hedgehog
+import qualified Data.Map as M
+
+import HW.Compiler
+import HW.StackMachine
+import Expr
+import HW.Eval
+import State.MyState
+
+genInt :: Gen Int
+genInt = Gen.int (Range.constant 0 100)
+
+genVar :: Gen String
+genVar = Gen.element ["x", "y", "z", "a", "b"]
+
+genNumExpr :: Gen (Expr String)
+genNumExpr = Expr.Num <$> genInt
+
+genVarExpr :: Gen (Expr String)
+genVarExpr = Expr.Var <$> genVar
+
+genAddExpr :: Gen (Expr String)
+genAddExpr = Expr.Plus <$> genExpr <*> genExpr
+
+genLetExpr :: Gen (Expr String)
+genLetExpr = Let <$> genVar <*> genExpr <*> genExpr
+
+genExpr :: Gen (Expr String)
+genExpr = Gen.recursive Gen.choice
+    [genNumExpr, genVarExpr]
+    [genAddExpr, genLetExpr]
+
+-- Checks for the proper execution of a stack program
+properExecution :: Property
+properExecution = property $ do
+    expr <- forAll (genExpr)
+    let program = compile expr
+    let initial = MachineState [] (M.fromList [("x", 10), ("y", 20), ("z", 30), ("a", 8), ("b", 33)]) 
+    let result = execProgram program initial
+    case result of
+        Right executed -> do
+            let stack = getStack executed
+            case stack of
+                [one] -> success
```

Comment:
You should also check that the result is correct.

Reply:
fixed